feat: PR for Code Review with Question 2

GHeeJeon opened the pull request.

GHeeJeon committed:
pr-apply advice

main 에 있던 게임 동작 로직 및 로그 출력을 각각 Judge 와 Recorder 분리
상수명 변경
플레이어 다이스 장착

GHeeJeon committed:
pr-apply fraud class with inheritance, unify indentation

NullPointerException 발생, 코드리뷰필요

GHeeJeon commented on DiceGame/src/player/FraudPlayer.java:
```diff
@@ -5,26 +5,34 @@
 @Setter
 
-public class FraudPlayer {
+public class FraudPlayer extends Player {
```
Player 클래스를 상속받도록 하여 코드를 작성했는데  Judge.java 에서 에러가 뜹니다.(Judge.java 의 에러가 뜨는 부분에 자세히 적어두었어요!)

GHeeJeon commented on DiceGame/src/judge/Judge.java:
```diff
@@ -6,25 +6,44 @@
+            player1.setTotal(player1.getTotal() + player1.getDice().roll());
+            recorder.recordToShowScore(player1);
+
+            recorder.recordToThrowDice(player2);
```
이 부분에서 NullPointerException 이 발생합니다.
<img width="711" alt="스크린샷 2022-07-28 오후 11 44 01" src="https://user-images.githubusercontent.com/47032054/181569383-0294c969-787f-44a2-9d6f-72cb4cabd673.png">
문제가 발생한 부분의 디버깅 결과입니다!

player2, 즉 사기 플레이어의 이름, 초기화된 토탈값, 장착한 다이스(사기 주사위) 가 잘 들어가있는데,
사기 플레이어가 일반 플레이어를 상속하는 과정에서 문제가 생겼는지 사기 플레이어(player2) 객체 내에서 일반 플레이어의 인스턴스를 참조하려고 해 NullPointerException 이 발생합니다. 어떻게 해결하면 좋을까요?

GHeeJeon commented on DiceGame/src/judge/Judge.java:
```diff
@@ -6,25 +6,44 @@
+            recorder.recordToShowScore(player1);
+
+            recorder.recordToThrowDice(player2);
+            player2.setTotal(player2.getTotal() + player2.getDice().roll());
```
이 부분에서 NullPointerException 이 발생합니다.
<img width="711" alt="스크린샷 2022-07-28 오후 11 44 01" src="https://user-images.githubusercontent.com/47032054/181569383-0294c969-787f-44a2-9d6f-72cb4cabd673.png">
문제가 발생한 부분의 디버깅 결과입니다!

player2, 즉 사기 플레이어의 이름, 초기화된 토탈값, 장착한 다이스(사기 주사위) 가 잘 들어가있는데,
사기 플레이어가 일반 플레이어를 상속받는 과정에서 문제가 생겼는지 사기 플레이어(player2) 객체 내에서 일반 플레이어의 인스턴스를 참조하려고 해 NullPointerException 이 발생합니다. 어떻게 해결하면 좋을까요?

Pokbab replied:
상속의 개념에 대해서 이해를 해야할 것 같아요
상속을 받으면 부모클래스의 모든걸 사용할 수 있죠. 멤버변수, 메서드
단, private으로 되어있는건 상속을 받아도 사용할 수 없어요. 꽁꽁 숨겨있는거죠

Player를 보니 dice가 private으로 지정되어 있습니다

그리고 FraudPlayer는 fraudDice를 갖고있어요.
```player2.getDice().roll()```에서 에러가 발생하는건, FraudPlayer에서 fraudDice가 아닌 부모클래스의 dice를 가져오려고 해서 에러가 나고 있는거에요.

코드를 아래와 같이 수정해보죠

``` java
public class Player {
    protected Dice dice; // protected로 변경하여 자식클래스도 사용할 수 있도록 수정
```
``` java
public class FraudPlayer extends Player {

    private FraudDice fraudDice; // 삭제합니다

    public FraudPlayer(String name, FraudDice fraudDice) {
        this.name = name;
        System.out.println("사기꾼 등장! 내 이름은 " + name);
        System.out.println("나는 사기 주사위를 가지고 있지!");
        this.fraudDice = fraudDice; // 삭제하고 아래와 같이 수정해보세요
        super.dice = fraudDice;
    }
```

Pokbab commented on DiceGame/src/dice/FraudDice.java:
```diff
@@ -9,17 +9,26 @@
-    public FraudDice(){
-        System.out.println("사기 주사위 생성 완료!");
+    public enum Level {
+        EASY, HARD, DEFAULT
```
👍

Pokbab commented on DiceGame/src/player/FraudPlayer.java:
```diff
@@ -5,26 +5,34 @@
-        // (플레이어의 total > 사기 플레이어의 total) then 사기 주사위 rollHard()
-        // (사기 플레이어의 total - 플레이어의 total > 6) then 사기 주사위 rollEasy()
-        // else 사기 주사위 노말모드
+    public int levelOfDice(Player player1, FraudPlayer player2) {
```
메서드는 객체(인스턴스)의 행위를 나타냅니다. 따라서 명사가 아닌 동사로 시작하는 단어를 주로 사용해요

그리고, 지금 이 메서드는 사기플레이어의 행동을 정의하는거니까
인자값으로 플레이어를 받지 말고, 주사위 값만 전달받아도 될 것 같아요~ 굳이 플레이어까지 끌고와서 getTotal()을 매번 호출할 필요 없는거죠.

그리고 주사위의 레벨을 정하는것도
``` java
FraudDice fraudDice = super.dice;
fraudDice.setLevel(레벨 enum);
```
이렇게 지정해주면 되겠네요

GHeeJeon replied:
감사합니다!
인자 부분-일반 플레이어 객체는 현재 주사위 합계값(int) 으로 바꾸었는데,
사기 플레이어는 쉽지 않네요ㅠㅠ
여러 파일에 변경사항이 생겨 새 pr 에 이어서 질문드리겠습니다!

GHeeJeon committed:
pr-apply advice

메서드는 동사로 시작할 것
접근제어자 protected 로 변경
setLevel() 파라미터를 주사위 합계로 바로 받을 것(일단 일반 플레이어만)

GHeeJeon commented:
감사합니다!
메서드 이름을 동사로 시작하도록 변경하고
상속을 위해 접근제어자 protected 로 변경했으며
setLevel() 파라미터를 주사위 합계(int)로 바로 받게끔 바꿔주었습니다.(일단 일반 플레이어만)
주사위의 레벨을 정하는 부분에 대해 이어서 새 pr 올리겠습니다!
